### PR TITLE
Desedification

### PR DESCRIPTION
**Pull request by stedolan**
The runtime and compiler must agree on various magic numbers (opcodes, exception IDs, and so on). There are currently a few mechanisms for doing this, none of which is satisfying:
- opcodes, opcode names, and the interpreter's jump table are generated by regexing the C source of `instruct.h` using `sed`.
- runtimedef.ml is generated using `sed` to process the C source of `fail.h` and the list of primitives.
- `config.ml` is generated from `config.mlp` using `sed`, `Makefile`, and some numbers that are manually kept in sync with `byterun/config.h`

In this pull request, lists like the list of exceptions are stored in [.tbl files](https://github.com/stedolan/ocaml/blob/3db30c2560510a349d3da8e2e4294ee87bf4763a/byterun/exceptions.tbl), which are directly `#include`'d into the [C source of the runtime](https://github.com/stedolan/ocaml/blob/3db30c2560510a349d3da8e2e4294ee87bf4763a/byterun/fail.h#L28-L32) (a technique sometimes known as [X-macros](http://en.wikipedia.org/wiki/X_Macro)). Similarly, the C preprocesor is run on some [OCaml code](https://github.com/stedolan/ocaml/blob/3db30c2560510a349d3da8e2e4294ee87bf4763a/bytecomp/runtimedef.ml.c) to expose these numbers to the compiler. `config.ml` is also generated using the C preprocessor, so it can [directly refer to constants](https://github.com/stedolan/ocaml/blob/f3f420bc7ab2cfaa5749ff3e388ca7b5be9c6b91/utils/config.ml.c#L65) from `byterun/config.h`, without needing numbers to be manually kept in sync.

The impetus for this is that in the multicore branch I have a few more magic numbers that the compiler and runtime must agree on, and I couldn't really stand adding yet more sed.

This pull request is against 4.02, since that's the most recent code common to multicore and trunk, but can be rebased to whatever.

### CODE REVIEW

**Comment by hhugo** [comment]
:+1:

**Comment by planar** [comment]
IIRC, we used `cpp` for such purposes way back in Caml Light times, and we found out that C preprocessors are not too consistent across platforms, especially when dealing with OCaml code.

So make sure it works under Windows.

Or maybe we should use `m4` for this.

File: config/auto-aux/cpp.c [hunk 1]
```diff
@@ -0,0 +1,6 @@
+/* This file should preprocess to the string "ok", modulo whitespace */
```
Comment (planar) [inline config/auto-aux/cpp.c]: What about preprocessors that insert `#"file" line` directives?

**Comment by xavierleroy** [comment]
Yes, running cpp on OCaml code is dicey, because OCaml code is not lexically-correct C (think e.g. of single quotes in identifiers), so different C preprocessors can do different things.  Now, you may be lucky if your OCaml code is trivial (e.g. "let foo = FOO").

I'd rather avoid m4, thank you very much.

File: config/auto-aux/cpp.c (hunk 1 above)
Comment (stedolan) [inline config/auto-aux/cpp.c]: That's what this check is for. `configure` tries running this through `cc -E -P`, to see if it produces `ok`.

Removing `#line` and `#file` with `-P` isn't specified by POSIX, but seems to be widely available. I've tested gcc and clang, and a quick look at docs for sun's cc and AIX cc say it exists (but I haven't run them). MSVC has this option under the name `/EP`, apparently.

**Comment by stedolan** [comment]
> Yes, running cpp on OCaml code is dicey, because OCaml code is not lexically-correct C (think e.g. 
> of  single quotes in identifiers), so different C preprocessors can do different things. Now, you may be
> lucky if your OCaml code is trivial (e.g. "let foo = FOO").

Yes. See [this change to a comment in config.ml](https://github.com/stedolan/ocaml/commit/1dc79316f19b6012d2a083b64c91eb2cc80d8d03#diff-0d331e832576e6edff71ea69d4bd0876L68) for an example (CPP disliked an apostrophe inside an OCaml comment).

I don't think CPP is a good general-purpose preprocessor for OCaml, but I think CPP on simple files is a reasonable way of getting magic numbers to coincide between C and OCaml. It is at any rate less dicey than running sed on C.

**Comment by bluddy** [comment]
Looks great. Reduces duplication, and one less unix-ism, too.

**Comment by xavierleroy** [comment]
Damien Doligez and I rediscussed this PR, and we agreed that the C preprocessor cannot be used safely to preprocess OCaml source files.  This introduces more potential problems than it solves.  I move to close this PR.

**Comment by xavierleroy** [comment]
In the absence of new information to support this PR, I'm closing it.

**Pull request closed without merging**